src/dune_pkg/opam_solver.ml

```diff
@@ -860,6 +880,33 @@ let solve_lock_dir
                       (Package_name.to_string name)
                       (Package_name.to_string dep_name)
                   ]));
+        let reachable =
+          let roots = Package_name.Map.keys local_packages in
+          let pkgs_by_name =
```

Comment:
The `solve_lock_dir` function is getting too long. Can you move this logic for computing the name->deps map to a separate function or move it into the `reachable` function itself.